admin/projects.go

```diff
@@ -207,6 +205,88 @@ func (s *Service) UpdateProject(ctx context.Context, proj *database.Project, opt
 	return proj, nil
 }
 
+// UpdateProjectVariables updates a project's variables and runs reconcile on the deployments.
+func (s *Service) UpdateProjectVariables(ctx context.Context, project *database.Project, environment string, vars map[string]string, unsetVars []string, userID string) ([]*database.ProjectVariable, error) {
+	if len(vars) == 0 && len(unsetVars) == 0 {
+		return nil, nil
+	}
+	txCtx, tx, err := s.DB.NewTx(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// Upsert variables
+	var updatedVars []*database.ProjectVariable
+	if len(vars) > 0 {
+		updatedVars, err = s.DB.UpsertProjectVariable(txCtx, project.ID, environment, vars, userID)
+		if err != nil {
+			_ = tx.Rollback()
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	// Delete unset variables
+	if len(unsetVars) > 0 {
+		err = s.DB.DeleteProjectVariables(txCtx, project.ID, environment, unsetVars)
+		if err != nil {
+			_ = tx.Rollback()
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	// Commit transaction
+	err = tx.Commit()
+	if err != nil {
+		_ = tx.Rollback()
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// Update deployments
+	s.Logger.Info("update project variables: updating deployments", observability.ZapCtx(ctx))
+
+	org, err := s.DB.FindOrganization(ctx, project.OrganizationID)
+	if err != nil {
+		return nil, err
+	}
+
+	annotations := s.NewDeploymentAnnotations(org, project)
+
+	ds, err := s.DB.FindDeploymentsForProject(ctx, project.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	vars, err = s.ResolveProdVariables(ctx, project.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(vars) == 0 {
+		// edge case : no prod variables to set (variable was deleted)
+		// but the runtime does not update variables if the new map is empty
+		// so we need to set a dummy variable to trigger the update
+		vars = map[string]string{"nonce": time.Now().Format(time.RFC3339Nano)}
+	}
```

Comment:
`nonce` is too likely to collide with a user-provided variable in the long run. Maybe `rill.internal.nonce`? 

Also, shouldn't this be set in `ResolveProdVariables` for consistent handling?

Reply:
No this is only set when user is trying to remove the last set `variable`(This is just a side-effect on how the runtime API works). 
In other cases it will not be set and that's why it is not required in `ResolveProdVariables`. 
We can change it to `rill.internal.nonce` but it is better to only set this when it is required.

Reply:
Okay, I'm just thinking about future cases where variables may be resolved and saved from other code. For example a ClickHouse provisioner might remove variables when being disabled. 

Maybe have a `forWriting bool` arg on `ResolveVariables` that adds the nonce field if there aren't any variables? Just so the code for resolving variables is actually centralized in `ResolveVariables`.

Reply:
I don't know how clickhouse provisioner will update variables but I think whenever variables are updated ideally it should call `UpdateProjectVariables`.
But yeah I am good with keeping it in `ResolveVariables` as well.